### PR TITLE
Human-in-the-Loop Implementation

### PR DESCRIPTION
**Pull request by mjschock**
Resolves https://github.com/openai/openai-agents-python/issues/636.

See https://github.com/openai/openai-agents-python/issues/636#issuecomment-3475335866.

```bash
OPENAI_API_KEY="your_api_key_here" uv run python examples/agent_patterns/human_in_the_loop.py

================================================================================
Run interrupted - tool approval required
================================================================================
State saved to result.json
Loading state from result.json

Tool call details:
  Agent: Weather Assistant
  Tool: get_temperature
  Arguments: {"city":"Oakland"}

Do you approve this tool call? (y/n): y
✓ Approved: get_temperature

Resuming agent execution...

================================================================================
Final Output:
================================================================================
The weather in Oakland is sunny, and the temperature is 20°C.
```

```bash
OPENAI_API_KEY="your_api_key_here" uv run python examples/agent_patterns/human_in_the_loop_stream.py 

================================================================================
Human-in-the-loop: approval required for the following tool calls:
================================================================================

Tool call details:
  Agent: Weather Assistant
  Tool: get_temperature
  Arguments: {"city":"Oakland"}

Do you approve this tool call? (y/n): y
✓ Approved: get_temperature

Resuming agent execution...

================================================================================
Final Output:
================================================================================
The current weather in Oakland is sunny, with a temperature of 20°C.

Done!
```

### CODE REVIEW

File: src/agents/run.py [hunk 1]
```diff
@@ -973,6 +1055,7 @@ async def _start_streaming(
         previous_response_id: str | None,
         conversation_id: str | None,
         session: Session | None,
+        run_state: RunState[TContext] | None = None,
     ):
         if streamed_result.trace:
             streamed_result.trace.start(mark_as_current=True)
```
[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/run.py:2244]: Fixed in a56ce0a. Added server conversation tracker priming at lines 1076-1079 to match the non-streaming implementation and prevent message duplication when resuming from RunState.

[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/run.py:2244]: This was already fixed in 74c50fd at line 914: `current_turn=run_state._current_turn if run_state else 0`. The turn counter is properly restored from the RunState.

**Comment by seratch** [comment]
Thanks for sending this patch! 

I currently don't have the bandwidth to check this in depth, but one thing I wanted to mention is that, while implementing the sessions feature in openai-agents-js project, I found that the internals of runner need to take various HITL patterns into consideration. There might not be necessary to make those changes in this Python SDK, but sufficient testing for the sessions scenarios is worth doing.

**Comment by mjschock** [comment]
> Thanks for sending this patch!
> 
> I currently don't have the bandwidth to check this in depth, but one thing I wanted to mention is that, while implementing the sessions feature in openai-agents-js project, I found that the internals of runner need to take various HITL patterns into consideration. There might not be necessary to make those changes in this Python SDK, but sufficient testing for the sessions scenarios is worth doing.

Happy to contribute! I added a couple examples using SQLiteSession and OpenAIConversationsSession and made sure they work:

```bash
OPENAI_API_KEY="your_api_key_here" uv run python examples/memory/memory_session_hitl_example.py 
=== Memory Session + HITL Example ===
Session id: :memory:
Enter a message to chat with the agent. Submit an empty line to exit.
The agent will ask for approval before using tools.

You: What cities does the Bay Bridge connect?
Assistant: The Bay Bridge connects San Francisco and Oakland in California.

You: What's the weather in those cities?

Agent HITL Assistant wants to call 'get_weather' with {"location":"San Francisco, CA"}. Approve? (y/n): y
Approved tool call.

Agent HITL Assistant wants to call 'get_weather' with {"location":"Oakland, CA"}. Approve? (y/n): y
Approved tool call.
Assistant: San Francisco is currently foggy with a temperature of 58°F. Oakland is sunny with a temperature of 72°F.

You: 
```

```bash
OPENAI_API_KEY="your_api_key_here" uv run python examples/memory/openai_session_hitl_example.py 
=== OpenAI Session + HITL Example ===
Enter a message to chat with the agent. Submit an empty line to exit.
The agent will ask for approval before using tools.

You: What cities does the Bay Bridge connect?
Assistant: The Bay Bridge, officially known as the San Francisco–Oakland Bay Bridge, connects the cities of **San Francisco** and **Oakland** in California.

You: What's the weather in those cities?

Agent HITL Assistant wants to call 'get_weather' with {"location":"San Francisco, CA"}. Approve? (y/n): y
Approved tool call.

Agent HITL Assistant wants to call 'get_weather' with {"location":"Oakland, CA"}. Approve? (y/n): y
Approved tool call.
Assistant: San Francisco is currently foggy and 58°F, while Oakland is sunny and 72°F.

You: 
```

I'm hoping that just about covers everything but lemme know if there are other areas I should make sure address.

**Comment by mjschock** [comment]
Thanks @seratch for adding it to the 0.6.x milestone! I'll make sure to address any feedback.

**Comment by mjschock** [comment]
Just a heads up. I noticed there are some discrepancies with the serialized RunState when interrupting in Python and resuming in TypeScript and visa versa. I'm addressing those this weekend and will push the update.

**Comment by seratch** [comment]
@codex Can you do reviews on the code changes in this PR with fresh eyes? Please focus on the potential issues for existing apps and edge case scenarios rather than general feedback.

File: src/agents/run.py [hunk 2]
```diff
@@ -1003,6 +1286,21 @@ async def _start_streaming(
 
             await AgentRunner._save_result_to_session(session, starting_input, [])
```
[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/run.py]: :+1: Fixed in https://github.com/openai/openai-agents-python/pull/2021/commits/466c306e8669a3f28c875ea6190371c2bb35438b

**Comment by seratch** [comment]
Thanks for working on this. I ran codex reviews many times and it detected these issues:

  - [P1] Resumed HITL flow never executes approved non-function tools — src/agents/_run_impl.py:626-716
    When resuming an interrupted turn after a tool approval, resolve_interrupted_turn only re-runs function tools and computer actions; it never executes queued shell, local
    shell, or apply_patch calls stored in processed_response (execute_shell_calls/execute_apply_patch_calls are not invoked). Any ShellTool/ApplyPatchTool configured with
    needs_approval will therefore return a ToolApprovalItem, but after the user approves and re-runs the agent the tool output is never produced, leaving the workflow stuck
    without the requested side effect or tool output. This affects any HITL approval scenario for these tools.

  - [P1] Resuming pending MCP approvals raises TypeError — src/agents/_run_impl.py:733-784
    When resuming a turn that contains hosted MCP approval requests, resolve_interrupted_turn stores the pending ToolApprovalItem objects in a set (pending_hosted_mcp_approvals).
    ToolApprovalItem is not hashable, so the first add() at this point raises TypeError: unhashable type: 'ToolApprovalItem', aborting the resume flow for any run interrupted on
    an MCP approval. This breaks HITL resumption whenever a hosted MCP approval is still pending.

  - [P1] Route local shell calls to remote shell tool — src/agents/_run_impl.py:1044-1052
    When processing model output, LocalShellCall items are dispatched to shell_tool if one exists (lines 1044–1050) before falling back to local_shell_tool. This means that if
    an agent registers both ShellTool and LocalShellTool, or if the model emits a local_shell_call despite only intending local execution, the call will be run through the remote
    shell tool instead of the local one, bypassing the local tool’s approval/on_approval hooks and executing in the wrong environment. Local shell calls should be handled by
    LocalShellTool; prioritizing shell_tool here misroutes the execution and can break HITL approval or shell behavior for any agent that enables both tools.

  - [P1] Streaming drops prior turn items — src/agents/run.py:3618-3624
    In _run_single_turn_streamed the pre_step_items list was changed to always be empty, so SingleStepResult.generated_items now contains only the current turn’s items instead
    of the accumulated history (streamed_result.new_items). On any streaming run with multiple turns, previous tool calls/outputs are dropped from both the next model request and
    the items persisted/emitted after the first turn, breaking conversation continuity and session persistence. This regression occurs whenever a streamed run progresses beyond
    the first turn.

  - [P1] Preserve max_turns when resuming from RunResult state — src/agents/result.py:213-223
    RunResult.to_state hardcodes max_turns=10 when creating a RunState, assuming the runner will override it. However Runner.run does not pull max_turns from the state—
    it uses the value passed by the caller or defaults to 10. If a run configured with a higher max_turns (e.g., 20) is interrupted for tool approval and resumed via state =
    result.to_state(); Runner.run(agent, state) without re‑passing max_turns, the resume path will treat current_turn from the state against a max_turns of 10 and will raise
    MaxTurnsExceeded on the next turn, even though budget remains. The max_turns used for the run needs to be persisted in the RunState instead of being reset to 10.

  - [P1] Deserialize only function approvals, breaks HITL for other tools — src/agents/run_state.py:983-999
    When restoring a run from JSON, the interruption reconstruction assumes every approval is a ResponseFunctionToolCall and blindly instantiates that type for each rawItem. Any
    HITL approval for shell, apply_patch, MCP, etc. will either fail validation or raise because their rawItem payloads do not match ResponseFunctionToolCall, making resuming
    from saved state impossible for those tool types. The deserializer needs to branch on the tool call type (or keep the original dict) instead of always coercing to a function
    call.

  - [P1] Deserializing interruptions assumes function tool calls — src/agents/run_state.py:1168-1177
    Resuming a saved RunState fails for approvals on non-function tools because from_json blindly instantiates ResponseFunctionToolCall for every interruption. If the pending
    approval was for a shell/apply_patch/hosted tool call (all supported by the new HITL flow), this constructor raises validation errors and the run cannot be resumed. Please
    deserialize rawItem without forcing it to a function call or branch by type so non-function approvals survive round‑tripping.

  - [P2] Avoid error-level logging on tracker creation — src/agents/run.py:161-168
    _ServerConversationTracker.post_init logs an error with a full stack trace every time a tracker is instantiated (lines 161‑168). This runs for all server-managed
    conversations, so normal runs now emit error-level messages even when nothing is wrong, polluting logs and alerting systems. Consider removing or downgrading this log, or
    guarding it behind a debug flag.

**Comment by mjschock** [comment]
> Thanks for working on this. I ran codex reviews many times and it detected these issues:
> 
>     * [P1] Resumed HITL flow never executes approved non-function tools — src/agents/_run_impl.py:626-716
>       When resuming an interrupted turn after a tool approval, resolve_interrupted_turn only re-runs function tools and computer actions; it never executes queued shell, local
>       shell, or apply_patch calls stored in processed_response (execute_shell_calls/execute_apply_patch_calls are not invoked). Any ShellTool/ApplyPatchTool configured with
>       needs_approval will therefore return a ToolApprovalItem, but after the user approves and re-runs the agent the tool output is never produced, leaving the workflow stuck
>       without the requested side effect or tool output. This affects any HITL approval scenario for these tools.
> 
>     * [P1] Resuming pending MCP approvals raises TypeError — src/agents/_run_impl.py:733-784
>       When resuming a turn that contains hosted MCP approval requests, resolve_interrupted_turn stores the pending ToolApprovalItem objects in a set (pending_hosted_mcp_approvals).
>       ToolApprovalItem is not hashable, so the first add() at this point raises TypeError: unhashable type: 'ToolApprovalItem', aborting the resume flow for any run interrupted on
>       an MCP approval. This breaks HITL resumption whenever a hosted MCP approval is still pending.
> 
>     * [P1] Route local shell calls to remote shell tool — src/agents/_run_impl.py:1044-1052
>       When processing model output, LocalShellCall items are dispatched to shell_tool if one exists (lines 1044–1050) before falling back to local_shell_tool. This means that if
>       an agent registers both ShellTool and LocalShellTool, or if the model emits a local_shell_call despite only intending local execution, the call will be run through the remote
>       shell tool instead of the local one, bypassing the local tool’s approval/on_approval hooks and executing in the wrong environment. Local shell calls should be handled by
>       LocalShellTool; prioritizing shell_tool here misroutes the execution and can break HITL approval or shell behavior for any agent that enables both tools.
> 
>     * [P1] Streaming drops prior turn items — src/agents/run.py:3618-3624
>       In _run_single_turn_streamed the pre_step_items list was changed to always be empty, so SingleStepResult.generated_items now contains only the current turn’s items instead
>       of the accumulated history (streamed_result.new_items). On any streaming run with multiple turns, previous tool calls/outputs are dropped from both the next model request and
>       the items persisted/emitted after the first turn, breaking conversation continuity and session persistence. This regression occurs whenever a streamed run progresses beyond
>       the first turn.
> 
>     * [P1] Preserve max_turns when resuming from RunResult state — src/agents/result.py:213-223
>       RunResult.to_state hardcodes max_turns=10 when creating a RunState, assuming the runner will override it. However Runner.run does not pull max_turns from the state—
>       it uses the value passed by the caller or defaults to 10. If a run configured with a higher max_turns (e.g., 20) is interrupted for tool approval and resumed via state =
>       result.to_state(); Runner.run(agent, state) without re‑passing max_turns, the resume path will treat current_turn from the state against a max_turns of 10 and will raise
>       MaxTurnsExceeded on the next turn, even though budget remains. The max_turns used for the run needs to be persisted in the RunState instead of being reset to 10.
> 
>     * [P1] Deserialize only function approvals, breaks HITL for other tools — src/agents/run_state.py:983-999
>       When restoring a run from JSON, the interruption reconstruction assumes every approval is a ResponseFunctionToolCall and blindly instantiates that type for each rawItem. Any
>       HITL approval for shell, apply_patch, MCP, etc. will either fail validation or raise because their rawItem payloads do not match ResponseFunctionToolCall, making resuming
>       from saved state impossible for those tool types. The deserializer needs to branch on the tool call type (or keep the original dict) instead of always coercing to a function
>       call.
> 
>     * [P1] Deserializing interruptions assumes function tool calls — src/agents/run_state.py:1168-1177
>       Resuming a saved RunState fails for approvals on non-function tools because from_json blindly instantiates ResponseFunctionToolCall for every interruption. If the pending
>       approval was for a shell/apply_patch/hosted tool call (all supported by the new HITL flow), this constructor raises validation errors and the run cannot be resumed. Please
>       deserialize rawItem without forcing it to a function call or branch by type so non-function approvals survive round‑tripping.
> 
>     * [P2] Avoid error-level logging on tracker creation — src/agents/run.py:161-168
>       _ServerConversationTracker.post_init logs an error with a full stack trace every time a tracker is instantiated (lines 161‑168). This runs for all server-managed
>       conversations, so normal runs now emit error-level messages even when nothing is wrong, polluting logs and alerting systems. Consider removing or downgrading this log, or
>       guarding it behind a debug flag.

I appreciate the feedback. I'll add tests to replicate the P1 failing scenarios and then fix them.

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
@codex review

File: src/agents/result.py [hunk 3]
```diff
@@ -170,6 +188,56 @@ def _release_last_agent_reference(self) -> None:
...
+        )
+
+        # Populate the state with data from the result
+        state._generated_items = self.new_items
+        state._model_responses = self.raw_responses
+        state._input_guardrail_results = self.input_guardrail_results
+        state._output_guardrail_results = self.output_guardrail_results
+        state._last_processed_response = self._last_processed_response
+        state._current_turn_persisted_item_count = self._current_turn_persisted_item_count
+        state.set_tool_use_tracker_snapshot(self._tool_use_tracker_snapshot)
```
[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/result.py:236]: Fixed in https://github.com/openai/openai-agents-python/pull/2021/commits/d93d0469eee500c5ce39aa4effe61ee2be7317d5

File: src/agents/run.py [hunk 4]
```diff
@@ -920,12 +1991,45 @@ def run_streamed(
...
+            # the counter might be 0 or incorrect. In this case, all items in
+            # generated_items were already saved, so set the counter to the length
+            # of generated_items to prevent duplication. For Python-to-Python
+            # resumes, the counter should already be correct, so we use it as-is.
+            _current_turn_persisted_item_count=(
+                (
+                    len(run_state._generated_items)
+                    if run_state._generated_items
+                    else 0
+                    if run_state._current_turn_persisted_item_count == 0
```
[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/run.py:2061]: Fixed in https://github.com/openai/openai-agents-python/pull/2021/commits/3891cecc82eda43286f73c95747da205fc5c36f2

**Comment by mjschock** [comment]
@codex review

File: src/agents/run_state.py [hunk 5]
```diff
@@ -0,0 +1,1817 @@
...
+
+    def _convert_output_item_to_protocol(self, raw_item_dict: dict[str, Any]) -> dict[str, Any]:
+        """Convert API-format tool output items to protocol format."""
+        converted = dict(raw_item_dict)
+        call_id = cast(Optional[str], converted.get("call_id") or converted.get("callId"))
+
+        converted["type"] = "function_call_result"
+
+        if not converted.get("name"):
+            converted["name"] = self._lookup_function_name(call_id or "")
```
[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/run_state.py]: Fixed in https://github.com/openai/openai-agents-python/pull/2021/commits/c218592f20a2abde19f826d413e1b8f861125aec

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
Re: [Codex comment on persisting server conversation metadata in RunState](https://github.com/openai/openai-agents-python/pull/2021#discussion_r2625341662)

Thanks for identifying this issue. After reviewing the TypeScript implementation (`openai-agents-js`), I found that the TypeScript SDK has the same limitation—`RunState` does not persist `conversationId`, `previousResponseId`, or `autoPreviousResponseId` fields.

Adding these fields to the Python `RunState` serialization would break cross-SDK compatibility because:

1. The TypeScript `SerializedRunState` schema uses Zod's `z.object()` which rejects unknown fields by default
2. If Python serializes a `RunState` with these fields, TypeScript deserialization would fail with validation errors
3. We're maintaining compatibility with the TypeScript SDK's behavior

To maintain cross-SDK compatibility, users need to pass `conversation_id`, `previous_response_id`, and `auto_previous_response_id` as kwargs when resuming from a saved `RunState`, just as they do in the TypeScript SDK. This matches the current behavior in both SDKs.

If this functionality is needed, it would require coordinating changes across both SDKs to add these fields to the shared schema, which is outside the scope of this PR.

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
@codex review

**Comment by mjschock** [comment]
@codex review

File: src/agents/run_state.py [hunk 6]
```diff
@@ -0,0 +1,1827 @@
...
+                    normalized_items.append(item)
+            original_input_serialized = normalized_items
+
+        result = {
+            "$schemaVersion": CURRENT_SCHEMA_VERSION,
+            "currentTurn": self._current_turn,
+            "currentAgent": {
+                "name": self._current_agent.name,
+            },
+            "originalInput": original_input_serialized,
```
[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/run_state.py:376]: @codex see https://github.com/openai/openai-agents-python/pull/2021#issuecomment-3663338845

[the remark this answers is not in the document]
Comment (mjschock) [reply src/agents/run_state.py:376]: :-1: I already mentioned this behavior matches openai-agents-js in https://github.com/openai/openai-agents-python/pull/2021#issuecomment-3663338845

**Comment by seratch** [comment]
Thank you so much for maintaining this branch for a while! 🙇 I will start checking this in depth next week, but before that, I ran local codex reviews (/review) so many times and got these feedback from it. I've verified there is no definite duplicate in the list but some of them might be already mentioned in this pull request discussion. Hope this helps!

Also, from my experience for TS SDK, I do understand implementing this along with sessions compatibility is really complex. That said, it seems this PR has huge diff now and it could be challenging to review for me. If you see anything to simplify in this PR, refactoring the code would be appreciated as well.

---

- [P1] Server conversation tracker returns non-serializable run items — openai-agents-python/src/agents/run.py:456-492
  In _ServerConversationTracker.prepare_input the loop over generated_items now appends run_item.raw_item directly (lines 456‑491) instead of using run_item.to_input_item() as before. For run
  items whose raw_item is a Pydantic model (e.g., ResponseFunctionToolCall or ResponseOutputMessage), this returns a model instance rather than the dict payload the Responses API expects, so the
  next request will contain non‑JSON‑serializable objects and fail whenever a BaseModel output/tool call needs to be replayed to the server-managed conversation. The code should continue
  converting run items to input dictionaries before adding them to the request.

- [P1] Generated items are repeatedly resent to model — openai-agents-python/src/agents/run.py:431-492
  In _ServerConversationTracker.prepare_input generated items are appended to input_items but never recorded in sent_items as before, so subsequent calls will treat the same raw_item as unseen and
  resend it to the model/session. Any multi-turn or retried run that calls prepare_input more than once will duplicate previously sent tool outputs/messages, bloating history and potentially
  changing model behavior or causing server-side "duplicate item" errors. Add the sent-item bookkeeping when enqueuing generated items to avoid resending them.

- [P1] Server-managed conversation sends unnormalized tool outputs — openai-agents-python/src/agents/run.py:431-492
  When server-managed conversations are used (conversation_id/previous_response_id flows), _ServerConversationTracker.prepare_input now injects each generated run item’s raw_item directly into the
  next request. ToolCallOutputItem.to_input_item() used to strip fields the Responses API rejects (e.g., status, shell_output, protocol-only function_call_result names). Skipping this
  normalization means hosted tool outputs or function results with protocol fields are sent verbatim and will be rejected by the API, breaking any turn that includes local tool execution while
  using server-managed conversations.

- [P1] prepare_input ignores fingerprinted sent items — openai-agents-python/src/agents/run.py:431-490
  In _ServerConversationTracker.prepare_input we only skip generated items that match by object id or server-tracked ids, but we never consult sent_item_fingerprints even though prime_from_state
  and track_server_items populate it specifically to cover rehydrated copies. When resuming a server-managed conversation (or after deserializing state), the generated items are new objects with
  different ids, so prepare_input will re-send the entire prior history/tool outputs back to the Responses API, duplicating messages and tool outputs. We should also skip items whose serialized
  fingerprint is already in sent_item_fingerprints to avoid replaying previously delivered content after a resume.

- [P1] Interrupted runs lose model response for resumption — openai-agents-python/src/agents/run.py:1213-1244
  When a fresh run hits a NextStepInterruption, the returned RunResult is built without appending the current step’s model_response or setting _last_processed_response (lines 1213‑1244). Calling
  to_state() and attempting to resume a HITL flow then produces a RunState with empty _model_responses and no processed response, causing Runner.run to raise UserError("No model response found in
  previous state"). Any interruption on the first pass is therefore not resumable. The model response and processed response need to be captured on this path before returning.

- [P1] Duplicates already persisted session items when resuming — src/agents/run.py:4195-4203
  When saving conversation history, _save_result_to_session is intended to skip items that were already persisted in a prior partial save, but if _current_turn_persisted_item_count >=
  len(new_items) it sets new_run_items = list(new_items) rather than an empty list. In resumed HITL flows this resubmits items that were already saved earlier in the turn, leading to duplicated
  messages/tool outputs in the backing session (OpenAI Conversations or SQLite) and potentially “item already exists” errors. The early-return should skip saving when everything is already
  persisted instead of resaving the entire list.

- [P1] Approvals run after already executing other tools — openai-agents-python/src/agents/_run_impl.py:350-432
  In RunImpl.execute_tools_and_side_effects we start all tool executions with asyncio.gather before checking for ToolApprovalItem interruptions, so computer/shell/apply_patch calls in the same
  model response will still run even when a function tool needs approval. This defeats the HITL expectation that the turn pauses before any side effects occur; e.g., a response containing both a
  tool requiring approval and a computer action will still perform the computer action without approval. The interruption check should precede executing non-approved tool calls so that no side
  effects happen until the approval is granted.

- [P1] Pending approvals ignored when resuming interrupted turn — openai-agents-python/src/agents/_run_impl.py:678-754
  When resuming from a saved RunState, resolve_interrupted_turn executes the queued tools and appends their results to new_items, but it never checks whether any of those executions returned a
  ToolApprovalItem. If a tool still requires approval (e.g., the user didn’t approve it, or an on_approval callback leaves it pending), ShellAction/ApplyPatchAction/FunctionTool will return a
  ToolApprovalItem, yet this code proceeds to NextStepRunAgain without setting NextStepInterruption or populating interruptions. The caller receives no interruption and the next turn will filter
  out the approval items, so the HITL approval is silently dropped and the tool never runs. The interruption check needs to mirror the one in execute_tools_and_side_effects to return
  NextStepInterruption whenever any tool still needs approval.

- [P1] Resuming after approval replays already-run tools — openai-agents-python/src/agents/_run_impl.py:690-724
  When a turn is interrupted for tool approval, tools that did not require approval are executed before the interruption is raised. On resume, resolve_interrupted_turn runs all computer/shell/
  apply_patch calls again (execute_computer_actions, execute_shell_calls, etc.) without filtering out the ones already run, and then appends their outputs because object IDs differ. If a model
  response mixes an approval-gated function call with a shell/apply_patch action, resuming after approval will re-run that shell/patch a second time, duplicating side effects and history entries.
  These calls should be skipped on resume unless they were awaiting approval.

- [P1] Drop providerData when deserializing interrupts MCP approvals — openai-agents-python/src/agents/run_state.py:1471-1484
  RunState deserialization strips providerData from stored items via _normalize_field_names, so a serialized ToolApprovalItem for a hosted MCP tool loses the providerData.type marker required to
  detect MCP approval requests when resuming. As a result, if a run is interrupted for MCP approval, saved to JSON, and later restored, resolve_interrupted_turn never finds those pending MCP
  approvals and they will not be executed or surfaced to the user, blocking continuation of that flow.

**Comment by mjschock** [comment]
> Thank you so much for maintaining this branch for a while! 🙇 I will start checking this in depth next week, but before that, I ran local codex reviews (/review) so many times and got these feedback from it. I've verified there is no definite duplicate in the list but some of them might be already mentioned in this pull request discussion. Hope this helps!
> 
> Also, from my experience for TS SDK, I do understand implementing this along with sessions compatibility is really complex. That said, it seems this PR has huge diff now and it could be challenging to review for me. If you see anything to simplify in this PR, refactoring the code would be appreciated as well.
> 
>     * [P1] Server conversation tracker returns non-serializable run items — openai-agents-python/src/agents/run.py:456-492
>       In _ServerConversationTracker.prepare_input the loop over generated_items now appends run_item.raw_item directly (lines 456‑491) instead of using run_item.to_input_item() as before. For run
>       items whose raw_item is a Pydantic model (e.g., ResponseFunctionToolCall or ResponseOutputMessage), this returns a model instance rather than the dict payload the Responses API expects, so the
>       next request will contain non‑JSON‑serializable objects and fail whenever a BaseModel output/tool call needs to be replayed to the server-managed conversation. The code should continue
>       converting run items to input dictionaries before adding them to the request.
> 
>     * [P1] Generated items are repeatedly resent to model — openai-agents-python/src/agents/run.py:431-492
>       In _ServerConversationTracker.prepare_input generated items are appended to input_items but never recorded in sent_items as before, so subsequent calls will treat the same raw_item as unseen and
>       resend it to the model/session. Any multi-turn or retried run that calls prepare_input more than once will duplicate previously sent tool outputs/messages, bloating history and potentially
>       changing model behavior or causing server-side "duplicate item" errors. Add the sent-item bookkeeping when enqueuing generated items to avoid resending them.
> 
>     * [P1] Server-managed conversation sends unnormalized tool outputs — openai-agents-python/src/agents/run.py:431-492
>       When server-managed conversations are used (conversation_id/previous_response_id flows), _ServerConversationTracker.prepare_input now injects each generated run item’s raw_item directly into the
>       next request. ToolCallOutputItem.to_input_item() used to strip fields the Responses API rejects (e.g., status, shell_output, protocol-only function_call_result names). Skipping this
>       normalization means hosted tool outputs or function results with protocol fields are sent verbatim and will be rejected by the API, breaking any turn that includes local tool execution while
>       using server-managed conversations.
> 
>     * [P1] prepare_input ignores fingerprinted sent items — openai-agents-python/src/agents/run.py:431-490
>       In _ServerConversationTracker.prepare_input we only skip generated items that match by object id or server-tracked ids, but we never consult sent_item_fingerprints even though prime_from_state
>       and track_server_items populate it specifically to cover rehydrated copies. When resuming a server-managed conversation (or after deserializing state), the generated items are new objects with
>       different ids, so prepare_input will re-send the entire prior history/tool outputs back to the Responses API, duplicating messages and tool outputs. We should also skip items whose serialized
>       fingerprint is already in sent_item_fingerprints to avoid replaying previously delivered content after a resume.
> 
>     * [P1] Interrupted runs lose model response for resumption — openai-agents-python/src/agents/run.py:1213-1244
>       When a fresh run hits a NextStepInterruption, the returned RunResult is built without appending the current step’s model_response or setting _last_processed_response (lines 1213‑1244). Calling
>       to_state() and attempting to resume a HITL flow then produces a RunState with empty _model_responses and no processed response, causing Runner.run to raise UserError("No model response found in
>       previous state"). Any interruption on the first pass is therefore not resumable. The model response and processed response need to be captured on this path before returning.
> 
>     * [P1] Duplicates already persisted session items when resuming — src/agents/run.py:4195-4203
>       When saving conversation history, _save_result_to_session is intended to skip items that were already persisted in a prior partial save, but if _current_turn_persisted_item_count >=
>       len(new_items) it sets new_run_items = list(new_items) rather than an empty list. In resumed HITL flows this resubmits items that were already saved earlier in the turn, leading to duplicated
>       messages/tool outputs in the backing session (OpenAI Conversations or SQLite) and potentially “item already exists” errors. The early-return should skip saving when everything is already
>       persisted instead of resaving the entire list.
> 
>     * [P1] Approvals run after already executing other tools — openai-agents-python/src/agents/_run_impl.py:350-432
>       In RunImpl.execute_tools_and_side_effects we start all tool executions with asyncio.gather before checking for ToolApprovalItem interruptions, so computer/shell/apply_patch calls in the same
>       model response will still run even when a function tool needs approval. This defeats the HITL expectation that the turn pauses before any side effects occur; e.g., a response containing both a
>       tool requiring approval and a computer action will still perform the computer action without approval. The interruption check should precede executing non-approved tool calls so that no side
>       effects happen until the approval is granted.
> 
>     * [P1] Pending approvals ignored when resuming interrupted turn — openai-agents-python/src/agents/_run_impl.py:678-754
>       When resuming from a saved RunState, resolve_interrupted_turn executes the queued tools and appends their results to new_items, but it never checks whether any of those executions returned a
>       ToolApprovalItem. If a tool still requires approval (e.g., the user didn’t approve it, or an on_approval callback leaves it pending), ShellAction/ApplyPatchAction/FunctionTool will return a
>       ToolApprovalItem, yet this code proceeds to NextStepRunAgain without setting NextStepInterruption or populating interruptions. The caller receives no interruption and the next turn will filter
>       out the approval items, so the HITL approval is silently dropped and the tool never runs. The interruption check needs to mirror the one in execute_tools_and_side_effects to return
>       NextStepInterruption whenever any tool still needs approval.
> 
>     * [P1] Resuming after approval replays already-run tools — openai-agents-python/src/agents/_run_impl.py:690-724
>       When a turn is interrupted for tool approval, tools that did not require approval are executed before the interruption is raised. On resume, resolve_interrupted_turn runs all computer/shell/
>       apply_patch calls again (execute_computer_actions, execute_shell_calls, etc.) without filtering out the ones already run, and then appends their outputs because object IDs differ. If a model
>       response mixes an approval-gated function call with a shell/apply_patch action, resuming after approval will re-run that shell/patch a second time, duplicating side effects and history entries.
>       These calls should be skipped on resume unless they were awaiting approval.
> 
>     * [P1] Drop providerData when deserializing interrupts MCP approvals — openai-agents-python/src/agents/run_state.py:1471-1484
>       RunState deserialization strips providerData from stored items via _normalize_field_names, so a serialized ToolApprovalItem for a hosted MCP tool loses the providerData.type marker required to
>       detect MCP approval requests when resuming. As a result, if a run is interrupted for MCP approval, saved to JSON, and later restored, resolve_interrupted_turn never finds those pending MCP
>       approvals and they will not be executed or surfaced to the user, blocking continuation of that flow.

I'm happy to contribute and I really appreciate you sending the edge cases my way. I've created tests that capture each failing edge case and the corresponding fixes. I'll spend some time this week to see if I can reduce the number of changes while maintaining the test cases. Less is more. =)

**Comment by mjschock** [comment]
@codex please take a fresh look at the full set of changes and do a thorough review

**Comment by mjschock** [comment]
@codex review

**Comment by seratch** [comment]
@mjschock Thank you so much for taking the time to work on this. Since yesterday, I've been working on the code based on your branch here. There are still some things to polish, and there are additional patterns that need to be supported (e.g., local shell tool migration and perhaps some others).

This enhancement is a fairly large change, and as a maintainer, I need to ensure the final quality. If you're okay with it, would it be all right if I take over your changes in a new pull request and continue collaborating with you on the remaining work? Of course, I'll credit your contribution using Co-authored-by, so your work will be fully preserved in the new PR. I'd really appreciate your understanding.

**Comment by mjschock** [comment]
> @mjschock Thank you so much for taking the time to work on this. Since yesterday, I've been working on the code based on your branch here. There are still some things to polish, and there are additional patterns that need to be supported (e.g., local shell tool migration and perhaps some others).
> 
> This enhancement is a fairly large change, and as a maintainer, I need to ensure the final quality. If you're okay with it, would it be all right if I take over your changes in a new pull request and continue collaborating with you on the remaining work? Of course, I'll credit your contribution using Co-authored-by, so your work will be fully preserved in the new PR. I'd really appreciate your understanding.

@seratch yep, np

**Comment by seratch** [comment]
Thank you so much again for building the foundation for HITL implementation! Let's continue working together at https://github.com/openai/openai-agents-python/pull/2230
